File: tests/cpp-tests/Classes/EffectsTest/EffectsTest.cpp
/****************************************************************************
 Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.
 
 http://www.cocos2d-x.org
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

#include "EffectsTest.h"
#include "../testResource.h"

USING_NS_CC;

EffectTests::EffectTests()
{
    ADD_TEST_CASE(Shaky3DDemo);
    ADD_TEST_CASE(Waves3DDemo);
    ADD_TEST_CASE(FlipX3DDemo);
    ADD_TEST_CASE(FlipY3DDemo);
    ADD_TEST_CASE(Lens3DDemo);
    ADD_TEST_CASE(Ripple3DDemo);
    ADD_TEST_CASE(LiquidDemo);
    ADD_TEST_CASE(WavesDemo);
    ADD_TEST_CASE(TwirlDemo);
    ADD_TEST_CASE(ShakyTiles3DDemo);
    ADD_TEST_CASE(ShatteredTiles3DDemo);
    ADD_TEST_CASE(ShuffleTilesDemo);
    ADD_TEST_CASE(FadeOutTRTilesDemo);
    ADD_TEST_CASE(FadeOutBLTilesDemo);
    ADD_TEST_CASE(FadeOutUpTilesDemo);
    ADD_TEST_CASE(FadeOutDownTilesDemo);
    ADD_TEST_CASE(TurnOffTilesDemo);
    ADD_TEST_CASE(WavesTiles3DDemo);
    ADD_TEST_CASE(JumpTiles3DDemo);
    ADD_TEST_CASE(SplitRowsDemo);
    ADD_TEST_CASE(SplitColsDemo);
    ADD_TEST_CASE(PageTurn3DDemo);
    ADD_TEST_CASE(PageTurn3DRectDemo);
}

Shaky3DDemo::Shaky3DDemo()
{
    _title = "Shaky3D";
    _subtitle = "";
}

cocos2d::ActionInterval* Shaky3DDemo::createEffect(float t)
{
    return Shaky3D::create(t, Size(15,10), 5, false);
}

Waves3DDemo::Waves3DDemo()
{
    _title = "Waves3D";
    _subtitle = "";
}

cocos2d::ActionInterval* Waves3DDemo::createEffect(float t)
{
    return Waves3D::create(t, Size(15,10), 5, 40);
}

cocos2d::ActionInterval* FlipX3DDemo::createEffect(float t)
{
    auto flipx  = FlipX3D::create(t);
    auto flipx_back = flipx->reverse();
    auto delay = DelayTime::create(2);
    return Sequence::create(flipx, delay, flipx_back, nullptr);
}

FlipX3DDemo::FlipX3DDemo()
{
    _title = "FlipX3D";
    _subtitle = "";
}

cocos2d::ActionInterval* FlipY3DDemo::createEffect(float t)
{
    auto flipy  = FlipY3D::create(t);
    auto flipy_back = flipy->reverse();
    auto delay = DelayTime::create(2);
    
    return Sequence::create(flipy, delay, flipy_back, nullptr);
}

FlipY3DDemo::FlipY3DDemo()
{
    _title = "FlipY3D";
    _subtitle = "";
}

cocos2d::ActionInterval* Lens3DDemo::createEffect(float t)
{
    auto size = Director::getInstance()->getWinSize();
    return Lens3D::create(t, Size(15,10), Vec2(size.width/2,size.height/2), 240);
}

Lens3DDemo::Lens3DDemo()
{
    _title = "Lens3D";
    _subtitle = "";
}

cocos2d::ActionInterval* Ripple3DDemo::createEffect(float t)
{
    auto size = Director::getInstance()->getWinSize();
    return Ripple3D::create(t, Size(32,24), Vec2(size.width/2,size.height/2), 240, 4, 160);
}

Ripple3DDemo::Ripple3DDemo()
{
    _title = "Ripple3D";
    _subtitle = "";
}

cocos2d::ActionInterval* LiquidDemo::createEffect(float t)
{
    return Liquid::create(t, Size(16,12), 4, 20);
}

LiquidDemo::LiquidDemo()
{
    _title = "Liquid";
    _subtitle = "";
}

cocos2d::ActionInterval* WavesDemo::createEffect(float t)
{
    return Waves::create(t, Size(16,12), 4, 20, true, true);
}

WavesDemo::WavesDemo()
{
    _title = "Waves";
    _subtitle = "";
}

cocos2d::ActionInterval* TwirlDemo::createEffect(float t)
{
    auto size = Director::getInstance()->getWinSize();
    return Twirl::create(t, Size(12,8), Vec2(size.width/2, size.height/2), 1, 2.5f);
}

TwirlDemo::TwirlDemo()
{
    _title = "Twirl";
    _subtitle = "";
}

cocos2d::ActionInterval* ShakyTiles3DDemo::createEffect(float t)
{
    return ShakyTiles3D::create(t, Size(16,12), 5, false) ;
}

ShakyTiles3DDemo::ShakyTiles3DDemo()
{
    _title = "ShakyTiles3D";
    _subtitle = "";
}

cocos2d::ActionInterval* ShatteredTiles3DDemo::createEffect(float t)
{
    return ShatteredTiles3D::create(t, Size(16,12), 5, false);
}

ShatteredTiles3DDemo::ShatteredTiles3DDemo()
{
    _title = "ShatteredTiles3D";
    _subtitle = "";
}

cocos2d::ActionInterval* ShuffleTilesDemo::createEffect(float t)
{
    auto shuffle = ShuffleTiles::create(t, Size(16,12), 25);
    auto shuffle_back = shuffle->reverse();
    auto delay = DelayTime::create(2);
    
    return Sequence::create(shuffle, delay, shuffle_back, nullptr);
}

ShuffleTilesDemo::ShuffleTilesDemo()
{
    _title = "ShuffleTiles";
    _subtitle = "";
}

cocos2d::ActionInterval* FadeOutTRTilesDemo::createEffect(float t)
{
    auto fadeout = FadeOutTRTiles::create(t, Size(16,12));
    auto back = fadeout->reverse();
    auto delay = DelayTime::create(0.5f);
    
    return Sequence::create(fadeout, delay, back, nullptr);
}

FadeOutTRTilesDemo::FadeOutTRTilesDemo()
{
    _title = "FadeOutTRTiles";
    _subtitle = "";
}

cocos2d::ActionInterval* FadeOutBLTilesDemo::createEffect(float t)
{
    auto fadeout = FadeOutBLTiles::create(t, Size(16,12));
    auto back = fadeout->reverse();
    auto delay = DelayTime::create(0.5f);
    
    return Sequence::create(fadeout, delay, back, nullptr);
}

FadeOutBLTilesDemo::FadeOutBLTilesDemo()
{
    _title = "FadeOutBLTiles";
    _subtitle = "";
}

cocos2d::ActionInterval* FadeOutUpTilesDemo::createEffect(float t)
{
    auto fadeout = FadeOutUpTiles::create(t, Size(16,12));
    auto back = fadeout->reverse();
    auto delay = DelayTime::create(0.5f);
    
    return Sequence::create(fadeout, delay, back, nullptr);
}

FadeOutUpTilesDemo::FadeOutUpTilesDemo()
{
    _title = "FadeOutUpTiles";
    _subtitle = "";
}

cocos2d::ActionInterval* FadeOutDownTilesDemo::createEffect(float t)
{
    auto fadeout = FadeOutDownTiles::create(t, Size(16,12));
    auto back = fadeout->reverse();
    auto delay = DelayTime::create(0.5f);
    
    return Sequence::create(fadeout, delay, back, nullptr);
}

FadeOutDownTilesDemo::FadeOutDownTilesDemo()
{
    _title = "FadeOutDownTiles";
    _subtitle = "";
}

cocos2d::ActionInterval* TurnOffTilesDemo::createEffect(float t)
{
    auto fadeout = TurnOffTiles::create(t, Size(48,32), 25);
    auto back = fadeout->reverse();
    auto delay = DelayTime::create(0.5f);
    
    return Sequence::create(fadeout, delay, back, nullptr);
}

TurnOffTilesDemo::TurnOffTilesDemo()
{
    _title = "TurnOffTiles";
    _subtitle = "";
}

cocos2d::ActionInterval* WavesTiles3DDemo::createEffect(float t)
{
    return WavesTiles3D::create(t, Size(15,10), 4, 120);
}

WavesTiles3DDemo::WavesTiles3DDemo()
{
    _title = "WavesTiles3D";
    _subtitle = "";
}

cocos2d::ActionInterval* JumpTiles3DDemo::createEffect(float t)
{
    return JumpTiles3D::create(t, Size(15,10), 2, 30);
}

JumpTiles3DDemo::JumpTiles3DDemo()
{
    _title = "JumpTiles3D";
    _subtitle = "";
}

cocos2d::ActionInterval* SplitRowsDemo::createEffect(float t)
{
    return SplitRows::create(t, 9);
}

SplitRowsDemo::SplitRowsDemo()
{
    _title = "SplitRows";
    _subtitle = "";
}

cocos2d::ActionInterval* SplitColsDemo::createEffect(float t)
{
    return SplitCols::create(t, 9);
}

SplitColsDemo::SplitColsDemo()
{
    _title = "SplitCols";
    _subtitle = "";
}

cocos2d::ActionInterval* PageTurn3DDemo::createEffect(float t)
{
    return PageTurn3D::create(t, Size(15,10));
}

PageTurn3DDemo::PageTurn3DDemo()
{
    _title = "PageTurn3D";
    _subtitle = "";
}

cocos2d::ActionInterval* PageTurn3DRectDemo::createEffect(float t)
{
    return PageTurn3D::create(t, Size(15,10));
}

PageTurn3DRectDemo::PageTurn3DRectDemo()
{
    _title = "PageTurn3D-Rect";
    _subtitle = "";
}

#define SID_RESTART        1

EffectBaseTest::EffectBaseTest()
: _gridNodeTarget(nullptr)
{
    
}

bool EffectBaseTest::init()
{
    if(TestCase::init())
    {
        LayerColor *background = LayerColor::create( Color4B(32,128,32,255) );
        this->addChild(background,-20);
        if(isRectEffect())
        {
            Size visibleSize = Director::getInstance()->getVisibleSize();
            Rect gridRect = Rect(visibleSize.width * 0.2,
                                 visibleSize.height * 0.2,
                                 visibleSize.width * 0.6,
                                 visibleSize.height * 0.6);
            _gridNodeTarget = NodeGrid::create(gridRect);
        }
        else
        {
            _gridNodeTarget = NodeGrid::create();
        }
        addChild(_gridNodeTarget, 0);
        _gridNodeTarget->runAction(createEffect(3));
        
        auto bg = Sprite::create(s_back3);
        _gridNodeTarget->addChild(bg, 0);
        bg->setPosition(VisibleRect::center());
        
        auto grossini = Sprite::create(s_pathSister2);
        _gridNodeTarget->addChild(grossini, 1);
        grossini->setPosition(VisibleRect::left().x+VisibleRect::getVisibleRect().size.width/3,VisibleRect::center().y);
        auto sc = ScaleBy::create(2, 5);
        auto sc_back = sc->reverse();
        grossini->runAction( RepeatForever::create(Sequence::create(sc, sc_back, nullptr) ) );
        
        auto tamara = Sprite::create(s_pathSister1);
        _gridNodeTarget->addChild(tamara, 1);
        tamara->setPosition(VisibleRect::left().x+2*VisibleRect::getVisibleRect().size.width/3,VisibleRect::center().y);
        auto sc2 = ScaleBy::create(2, 5);
        auto sc2_back = sc2->reverse();
        tamara->runAction( RepeatForever::create(Sequence::create(sc2, sc2_back, nullptr)) );
        
        schedule( CC_SCHEDULE_SELECTOR(EffectBaseTest::checkAnim) );
        
        return true;
    }
    return false;
}

void EffectBaseTest::checkAnim(float dt)
{
    if ( _gridNodeTarget->getNumberOfRunningActions() == 0 && _gridNodeTarget->getGrid() != nullptr)
        _gridNodeTarget->setGrid(nullptr);
}

EffectBaseTest::~EffectBaseTest(void)
{
}

File: tests/js-tests/src/EffectsTest/EffectsTest.js
/****************************************************************************
 Copyright (c) 2008-2010 Ricardo Quesada
 Copyright (c) 2011-2012 cocos2d-x.org
 Copyright (c) 2013-2014 Chukong Technologies Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/


var effectsTestSceneIdx = -1;

//
// Base Layer
//

var EffectsBaseLayer = BaseTestLayer.extend({
    code:function () {
        return "";
    },
    // callbacks
    onRestartCallback:function (sender) {
        var s = new EffectsTestScene();
        s.addChild(restartEffectsTest());
        director.runScene(s);
    },
    onNextCallback:function (sender) {
        var s = new EffectsTestScene();
        s.addChild(nextEffectsTest());
        //director.runScene(cc.TransitionZoomFlipX.create(5, s));
        director.runScene(s);
    },
    onBackCallback:function (sender) {
        var s = new EffectsTestScene();
        s.addChild(previousEffectsTest());
        director.runScene(s);
    },
    onEnter:function () {
        this._super();

        var node = new cc.Node();

        //Whether to demonstrate the effects inside a smaller rect
        var nodeGrid = new cc.NodeGrid();
        nodeGrid.addChild(node);
        nodeGrid.runAction(this.getEffect(3));
        this.addChild( nodeGrid );

        // back gradient
        var gradient = new cc.LayerGradient( cc.color(255,0,0,255), cc.color(255,255,0,255));
        node.addChild( gradient );

        // back image
        var bg = new cc.Sprite(s_back3);
        bg.x = winSize.width/2;
        bg.y = winSize.height/2;
        node.addChild( bg );

        var sister1 = new cc.Sprite(s_pathSister1);
        sister1.x = winSize.width/3;
        sister1.y = winSize.height/2;
        node.addChild( sister1, 1 );

        var sister2 = new cc.Sprite(s_pathSister2);
        sister2.x = winSize.width*2/3;
        sister2.y = winSize.height/2;
        node.addChild( sister2, 1 );

        var sc = cc.scaleBy(2, 5);
        var sc_back = sc.reverse();
        var seq = cc.sequence( sc, sc_back );
        var repeat = seq.repeatForever();

        sister1.runAction( repeat );
        sister2.runAction( repeat.clone() );
    },

    getEffect:function(duration) {
        // override me
        return cc.moveBy(2, cc.p(10,10) );
    },

    // automation
    numberOfPendingTests:function() {
        return ( (arrayOfEffectsTest.length-1) - effectsTestSceneIdx );
    },

    getTestNumber:function() {
        return effectsTestSceneIdx;
    }

});

//------------------------------------------------------------------
//
// Tests
//
//------------------------------------------------------------------
var Shaky3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "Shaky 3D";
    },
    code:function () {
        return "a = cc.shaky3D(duration, gridSize, range, shakeZ)";
    },
    getEffect:function(duration) {
        return cc.shaky3D( duration, cc.size(15,10), 5, false );
    }
});

var Waves3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "Waves 3D";
    },
    code:function () {
        return "a = cc.waves3D(duration, gridSize, range, shakeZ)";
    },
    getEffect:function(duration) {
        return cc.waves3D(duration, cc.size(15,10), 5, 40 );
    }
});

var FlipXTest = EffectsBaseLayer.extend({
    title:function () {
        return "FlipX3D";
    },
    code:function () {
        return "a = cc.flipX3D(duration )";
    },
    getEffect:function(duration) {
        var a = cc.flipX3D(duration);
        var delay = cc.delayTime(2);
        var r = a.reverse();
        return cc.sequence( a, delay, r );
    }
});

var FlipYTest = EffectsBaseLayer.extend({
    title:function () {
        return "FlipY3D";
    },
    code:function () {
        return "a = cc.flipY3D(duration )";
    },
    getEffect:function(duration) {
        var a = cc.flipY3D(duration );
        var delay = cc.delayTime(2);
        var r = a.reverse();
        return cc.sequence( a, delay, r );
    }
});

var Lens3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "Lens3D";
    },
    code:function () {
        return "a = cc.lens3D(duration, gridSize, position, radius)";
    },
    getEffect:function(duration) {
        return cc.lens3D( duration, cc.size(15,10), cc.p(winSize.width/2, winSize.height/2), 240);
    }
});

var Ripple3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "Ripple3D";
    },
    code:function () {
        return "a = cc.ripple3D(duration, gridSize, position, radius, waves, amplitude)";
    },
    getEffect:function(duration) {
        return cc.ripple3D( duration, cc.size(32,24), cc.p(winSize.width/2, winSize.height/2), 240, 4, 160);
    }
});

var LiquidTest = EffectsBaseLayer.extend({
    title:function () {
        return "Liquid";
    },
    code:function () {
        return "a = cc.liquid(duration, gridSize, waves, amplitude)";
    },
    getEffect:function(duration) {
        return cc.liquid( duration, cc.size(16,12), 4, 20);
    }
});

var WavesTest = EffectsBaseLayer.extend({
    title:function () {
        return "Waves";
    },
    code:function () {
        return "a = cc.waves(duration, gridSize, waves, amplitude, horizontal, vertical)";
    },
    getEffect:function(duration) {
        return cc.waves( duration, cc.size(16,12), 4, 20, true, true);
    }
});

var TwirlTest = EffectsBaseLayer.extend({
    title:function () {
        return "Twirl";
    },
    code:function () {
        return "a = cc.twirl(duration, gridSize, position, twirls, amplitude)";
    },
    getEffect:function(duration) {
        return cc.twirl( duration, cc.size(12,8), cc.p(winSize.width/2, winSize.height/2), 1, 2.5);
    }
});

var ShakyTiles3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "ShakyTiles3D";
    },
    code:function () {
        return "a = cc.shakyTiles3D(duration, gridSize, range, shakeZ)";
    },
    getEffect:function(duration) {
        return cc.shakyTiles3D( duration, cc.size(16,12), 5, false);
    }
});

var ShatteredTiles3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "ShatteredTiles3D";
    },
    code:function () {
        return "a = cc.shatteredTiles3D(duration, gridSize, range, shatterZ)";
    },
    getEffect:function(duration) {
        return cc.shatteredTiles3D( duration, cc.size(16,12), 5, false);
    }
});

var ShuffleTilesTest = EffectsBaseLayer.extend({
    title:function () {
        return "ShuffleTiles";
    },
    code:function () {
        return "a = cc.shuffleTiles(duration, gridSize, seed)";
    },
    getEffect:function(duration) {
        var action = cc.shuffleTiles( duration, cc.size(16,12), 25);
        var delay = cc.delayTime(2);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var FadeOutTRTilesTest = EffectsBaseLayer.extend({
    title:function () {
        return "FadeOutTRTilesTest";
    },
    code:function () {
        return "a = cc.fadeOutTRTiles(duration, gridSize)";
    },
    getEffect:function(duration) {
        var action = cc.fadeOutTRTiles( duration, cc.size(16,12));
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var FadeOutBLTilesTest = EffectsBaseLayer.extend({
    title:function () {
        return "FadeOutBLTilesTest";
    },
    code:function () {
        return "a = cc.fadeOutBLTiles(duration, gridSize)";
    },
    getEffect:function(duration) {
        var action = cc.fadeOutBLTiles( duration, cc.size(16,12));
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var FadeOutUpTilesTest = EffectsBaseLayer.extend({
    title:function () {
        return "FadeOutUpTilesTest";
    },
    code:function () {
        return "a = cc.fadeOutUpTiles(duration, gridSize)";
    },
    getEffect:function(duration) {
        var action = cc.fadeOutUpTiles( duration, cc.size(16,12));
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var FadeOutDownTilesTest = EffectsBaseLayer.extend({
    title:function () {
        return "FadeOutDownTilesTest";
    },
    code:function () {
        return "a = cc.fadeOutDownTiles(duration, gridSize)";
    },
    getEffect:function(duration) {
        var action = cc.fadeOutDownTiles( duration, cc.size(16,12));
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var TurnOffTilesTest = EffectsBaseLayer.extend({
    title:function () {
        return "TurnOffTiles";
    },
    code:function () {
        return "a = cc.turnOffTiles(duration, gridSize, seed)";
    },
    getEffect:function(duration) {
        var action = cc.turnOffTiles( duration, cc.size(48,32), 25);
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var WavesTiles3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "WavesTiles3D";
    },
    code:function () {
        return "a = cc.wavesTiles3D(duration, gridSize, waves, amplitude)";
    },
    getEffect:function(duration) {
        var action = cc.wavesTiles3D( duration, cc.size(16,12), 4, 120);
        return action;
    }
});


var JumpTiles3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "JumpTiles3D";
    },
    code:function () {
        return "a = cc.jumpTiles3D(duration, gridSize, jumps, amplitude)";
    },
    getEffect:function(duration) {
        var action = cc.jumpTiles3D(duration, cc.size(16,12), 2, 30);
        return action;
    }
});

var SplitRowsTest = EffectsBaseLayer.extend({
    title:function () {
        return "SplitRows";
    },
    code:function () {
        return "a = cc.splitRows(duration, rows)";
    },
    getEffect:function(duration) {
        var action = cc.splitRows(duration, 9);
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var SplitColsTest = EffectsBaseLayer.extend({
    title:function () {
        return "SplitCols";
    },
    code:function () {
        return "a = cc.splitCols(duration, cols)";
    },
    getEffect:function(duration) {
        var action = cc.splitCols(duration, 9);
        var delay = cc.delayTime(0.5);
        var back = action.reverse();
        var seq = cc.sequence( action, delay, back);
        return seq;
    }
});

var PageTurn3DTest = EffectsBaseLayer.extend({
    title:function () {
        return "PageTurn3D";
    },
    code:function () {
        return "a = cc.pageTurn3D(duration, gridSize)";
    },
    getEffect:function(duration) {
        var action = cc.pageTurn3D(duration, cc.size(15,10));
        return action;
    }
});

var PageTurn3DInRectTest = BaseTestLayer.extend({
    title:function () {
        return "PageTurn3DInRectTest";
    },
    code:function () {
        return "a = cc.pageTurn3D(duration, gridSize)";
    },
    // callbacks
    onRestartCallback:function (sender) {
        var s = new EffectsTestScene();
        s.addChild(restartEffectsTest());
        director.runScene(s);
    },
    onNextCallback:function (sender) {
        var s = new EffectsTestScene();
        s.addChild(nextEffectsTest());
        //director.runScene(cc.TransitionZoomFlipX.create(5, s));
        director.runScene(s);
    },
    onBackCallback:function (sender) {
        var s = new EffectsTestScene();
        s.addChild(previousEffectsTest());
        director.runScene(s);
    },
    onEnter:function(){
        this._super();

        //var node = new cc.Node();
        var visiableSize = director.getVisibleSize();
        var gridRect = cc.rect(visiableSize.width*0.1,
            visiableSize.height*0.1,
            visiableSize.width*0.4,
            visiableSize.height*0.4);
        var gridNodeTarget = new cc.NodeGrid(gridRect);

        gridNodeTarget.runAction(this.getEffect(3));
        this.addChild( gridNodeTarget );

        // back gradient
        var background = new cc.LayerGradient( cc.color(255,0,0,255), cc.color(255,255,0,255));
        gridNodeTarget.addChild( background );

        // back image
        var bg = new cc.Sprite(s_back3);
        bg.x = winSize.width/2;
        bg.y = winSize.height/2;
        gridNodeTarget.addChild( bg );

        var sister1 = new cc.Sprite(s_pathSister1);
        sister1.x = winSize.width/3;
        sister1.y = winSize.height/2;
        gridNodeTarget.addChild( sister1, 1 );

        var sister2 = new cc.Sprite(s_pathSister2);
        sister2.x = winSize.width*2/3;
        sister2.y = winSize.height/2;
        gridNodeTarget.addChild( sister2, 1 );

        var sc = cc.scaleBy(2, 5);
        var sc_back = sc.reverse();
        var seq = cc.sequence( sc, sc_back );
        var repeat = seq.repeatForever();

        sister1.runAction( repeat );
        sister2.runAction( repeat.clone() );
    },

    getEffect:function(duration) {
        var action = cc.pageTurn3D(duration, cc.size(15,10));
        return action;
    }
});

//
// Order of tests
//
var EffectsTestScene = TestScene.extend({
    runThisTest:function (num) {
        effectsTestSceneIdx = (num || num == 0) ? (num - 1) : -1;
        var layer = nextEffectsTest();
        this.addChild(layer);

        director.runScene(this);
    }
});

//
// Flow control
//
var arrayOfEffectsTest = [
    Shaky3DTest,
    Waves3DTest,
    FlipXTest,
    FlipYTest,
    Lens3DTest,
    Ripple3DTest,
    LiquidTest,
    WavesTest,
    TwirlTest,
    ShakyTiles3DTest,
    ShatteredTiles3DTest,
    ShuffleTilesTest,
    FadeOutTRTilesTest,
    FadeOutBLTilesTest,
    FadeOutUpTilesTest,
    FadeOutDownTilesTest,
    TurnOffTilesTest,
    WavesTiles3DTest,
    JumpTiles3DTest,
    SplitRowsTest,
    SplitColsTest,
    PageTurn3DTest,
    PageTurn3DInRectTest
];

var nextEffectsTest = function () {
    effectsTestSceneIdx++;
    effectsTestSceneIdx = effectsTestSceneIdx % arrayOfEffectsTest.length;

    if(window.sideIndexBar){
        effectsTestSceneIdx = window.sideIndexBar.changeTest(effectsTestSceneIdx, 14);
    }

    return new arrayOfEffectsTest[effectsTestSceneIdx]();
};
var previousEffectsTest = function () {
    effectsTestSceneIdx--;
    if (effectsTestSceneIdx < 0)
        effectsTestSceneIdx += arrayOfEffectsTest.length;

    if(window.sideIndexBar){
        effectsTestSceneIdx = window.sideIndexBar.changeTest(effectsTestSceneIdx, 14);
    }

    return new arrayOfEffectsTest[effectsTestSceneIdx]();
};
var restartEffectsTest = function () {
    return new arrayOfEffectsTest[effectsTestSceneIdx]();
};